src/util/prolog/prologStringManipulation.ts

```diff
@@ -1,41 +1,122 @@
 }
 
 function removeEmptyLines(prologProgram: string) {
+	const fetchEmptyLinesRegExp = /\n(?:\s)*\n/gm;
+
 	return prologProgram.replace(fetchEmptyLinesRegExp, "\n");
 }
 
 function removeDoubleWhitespaces(prologProgram: string) {
+	const fetchDoubleWhitespacesRegExp = /  */gm;
+
 	return prologProgram.replace(fetchDoubleWhitespacesRegExp, " ");
 }
 
 export function removeNonFunctionalities(prologProgram: string) {
 	return removeComments(removeEmptyLines(removeDoubleWhitespaces(prologProgram)));
 }
+
+export function fixVariableShadowingInImport(program: string, importedProgram: string): string {
+	const fetchImportedPredicatesRegExp = /:-[\r\t\f\v ]*module\([^,)]*,\[([^)]*)\][\r\t\f\v ]*\)[\r\t\f\v ]*/;
+
+	const useModuleSecondParam: string[] | null = fetchImportedPredicatesRegExp.exec(importedProgram);
+	if (useModuleSecondParam == null) {
+		return identifyPredicatesUniquely(program, importedProgram);
+	}
+	// The following regex will return in group one the name of the functions
+	const importedPredicates: string[] = fetchAllMatchesForAGroup(useModuleSecondParam![1], /[\r\t\f\v ]*(\w*)\/\d,?/g, 1);
+	return identifyPredicatesUniquely(program, importedProgram, importedPredicates);
+}
+
+function identifyPredicatesUniquely(program: string, importedProgram: string, importedPredicates?: string[]): string {
+
+	// imports shouldn't be renamed
+	let predicatesToNotRename: string[];
+	// Checking for undefined based on this recommendation:
+	// https://basarat.gitbooks.io/typescript/docs/javascript/null-undefined.html
+	if (importedPredicates == null) {
+		predicatesToNotRename = [];
+	} else {
+		predicatesToNotRename = importedPredicates;
+	}
+
+	// predicates which have to be renamed are the ones which occure in both the main and the imported program
+	const mainProgramPredicates: Set<string> = getNonBuiltInPredicates(program);
+	const predicatesToRename: Set<string> = getNonBuiltInPredicates(importedProgram);
+	const toRemove: Set<string> = new Set(); // Externalize deletions
+	predicatesToRename.forEach((predicate: string) => {
+		if (!mainProgramPredicates.has(predicate)) {
+			toRemove.add(predicate);
+		}
+	});
+	// Externalized Deletion
+	toRemove.forEach((predicate: string) => predicatesToRename.delete(predicate));
+	predicatesToNotRename.forEach((predicate: string) => predicatesToRename.delete(predicate));
+
+	// Renaming Process
+	predicatesToRename.forEach((toRename: string) => {
+		const getToRenameRegExp = new RegExp(toRename + "(\(.*?\))", "g");
+		importedProgram = importedProgram.replace(getToRenameRegExp, (_, parameters: string) => {
+
+			let identifier = 1;
+			while (mainProgramPredicates.has(toRename + "_" + identifier)) {
+				identifier++;
+			}
+
+			return toRename + "_" + identifier + parameters;
+
+		});
+	});
+
+	return importedProgram;
+}
+
+function getNonBuiltInPredicates(program: string): Set<string> {
+	// This doesn't take comments into account so in "%predicate()" would find predicate()
```

Comment:
Just curious: Is this how you want it to be in the end or is it your goal to change this later?

Reply:
I didn't plan to change it. It was more of a note to justify why I call `removeNonFunctionalities(program)` in line 114.

The part with the new line error can be avoided with a different regexp, but it would be more vunerable to change it and less efficient. The current one is not very efficient already.